router/throttler/throttler.go

```diff
@@ -1,173 +1,257 @@
 package throttler
 
 import (
+	"context"
 	"fmt"
 	"time"
 
+	"github.com/go-redis/redis/v9"
+
 	"github.com/rudderlabs/rudder-server/config"
-	"github.com/rudderlabs/rudder-server/router/throttler/ratelimiter"
+	"github.com/rudderlabs/rudder-server/internal/throttling"
 	"github.com/rudderlabs/rudder-server/utils/logger"
 )
 
 const (
-	DESTINATION_LEVEL = "destination"
-	USER_LEVEL        = "user"
-	ALL_LEVELS        = "all"
+	algoTypeGoRate         = "gorate"
+	algoTypeGCRA           = "gcra"
+	algoTypeRedisGCRA      = "redis-gcra"
+	algoTypeRedisSortedSet = "redis-sorted-set"
 )
 
-// Throttler is an interface for throttling functions
-type Throttler interface {
-	CheckLimitReached(destID, userID string, currentTime time.Time) bool
-	Inc(destID, userID string, currentTime time.Time)
-	Dec(destID, userID string, count int64, currentTime time.Time, atLevel string)
-	IsEnabled() bool
-	IsUserLevelEnabled() bool
-	IsDestLevelEnabled() bool
+type limiter interface {
+	// Limit should return true if the request can be done or false if the limit is reached.
+	Limit(ctx context.Context, cost, rate, window int64, key string) (
+		allowed bool, ret throttling.TokenReturner, err error,
+	)
 }
 
-type Limiter struct {
+type limiterSettings struct {
 	enabled     bool
-	eventLimit  int
+	eventLimit  int64
 	timeWindow  time.Duration
-	ratelimiter *ratelimiter.RateLimiter
+	rateLimiter limiter
 }
 
-type Settings struct {
-	limit               int
+type destinationSettings struct {
+	limit               int64
 	timeWindow          int64
-	userLevelLimit      int
+	userLevelLimit      int64
 	userLevelTimeWindow int64
 }
 
-// HandleT is a Handle for event limiter
-type HandleT struct {
-	destinationName string
-	destLimiter     *Limiter
-	userLimiter     *Limiter
+// Client is a Handle for event limiterSettings
+type Client struct {
+	algoType      string
+	redisClient   *redis.Client
```

Comment:
We don't need it but the alternative would be a map with some lock around it. Also the map would be a global variable to the `throttler` package. The `redisAddr` is unique enough to be used as the map key. Then if a client for that addr is already defined we use that instead of initializing a new one. Wdyt of this approach? Or if you have a better one in mind I'm all ears :+1: